### PR TITLE
feat(streaming): introduce streaming AsOf JOIN executor

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Introduce streaming AsOf join executor. https://duckdb.org/docs/guides/sql_features/asof_join.html

When a record comes from the left side, it will look up the first matched record on the right side.

When a record comes from the right side, we first find out which previously matched record will be affected, and then send delete/insert those records, and send insert/delete the new matched result.

We use a secondary index in `JoinEntryState` to do the range scan.

#17765

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: proto/stream_plan.proto [hunk 1]
```diff
@@ -455,6 +455,34 @@ message HashJoinNode {
...
+  repeated uint32 output_indices = 8;
+  // Left deduped input pk indices. The pk of the left_table and
+  // left_degree_table is  [left_join_key | left_deduped_input_pk_indices]
+  // and is expected to be the shortest key which starts with
+  // the join key and satisfies unique constrain.
+  repeated uint32 left_deduped_input_pk_indices = 9;
+  // Right deduped input pk indices. The pk of the right_table and
+  // right_degree_table is  [right_join_key | right_deduped_input_pk_indices]
+  // and is expected to be the shortest key which starts with
+  // the join key and satisfies unique constrain.
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply proto/stream_plan.proto:480]: Will fix the comment

File: src/stream/src/executor/asof_join.rs [hunk 2]
```diff
@@ -0,0 +1,1379 @@
...
+                                        .get_by_indexed_pk(smallest_pk, &side_update.all_data_types)
+                                    {
+                                        (
+                                            Some(Either::Left(to_delete_row?.row)),
+                                            Some(Either::Right(row)),
+                                        )
+                                    } else {
+                                        // Something wrong happened. Ignore this row in non strict consistency mode.
+                                        (None, None)
+                                    }
```
Comment (st1page) [inline src/stream/src/executor/asof_join.rs:795]: need to check with `enable_strict_consistency()`

File: src/stream/src/executor/asof_join.rs [hunk 3]
```diff
@@ -0,0 +1,1379 @@
...
+                                yield chunk;
+                            }
+                        } else if let Some(chunk) =
+                            hashjoin_chunk_builder.forward_if_not_matched(Op::Delete, row)
+                        {
+                            yield chunk;
+                        }
+                        side_update.ht.delete_row(key, row)?;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/asof_join.rs:694]: There are more diference here `with_match_on_insert/with_match_on_delete`. `ht.insert_row/ ht.delete_row`

File: src/stream/src/executor/asof_join.rs [hunk 4]
```diff
@@ -0,0 +1,1379 @@
...
+    // inequality_watermarks: &'a Watermark,
+    chunk: StreamChunk,
+    chunk_size: usize,
+    cnt_rows_received: &'a mut u32,
+    high_join_amplification_threshold: usize,
+}
+
+impl<K: HashKey, S: StateStore, const T: AsOfJoinTypePrimitive> AsOfJoinExecutor<K, S, T> {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/asof_join.rs:179]: There are a lot of subtle differences between building asof join and hash join. Extracting common logic costs more than it earns.

File: src/stream/src/executor/asof_join.rs [hunk 5]
```diff
@@ -0,0 +1,1379 @@
...
+                        side_update.ht.insert_row(key, row).await?;
+                    }
+                    Op::Delete | Op::UpdateDelete => {
+                        side_update.ht.delete_row(key, row)?;
+                    }
+                }
+            } else {
+                // Row which violates null-safe bitmap will never be matched so we need not
+                // store.
+                // Noop here because we only support left outer AsOf join.
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/asof_join.rs:938]: In the planner I'm working on.

File: src/stream/src/executor/asof_join.rs (hunk 2 above)
Comment (yuhao-su) [reply src/stream/src/executor/asof_join.rs:795]: Checked inside the `get_by_indexed_pk`

File: src/stream/src/executor/asof_join.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (st1page) [reply src/stream/src/executor/asof_join.rs:694]: I just mean that change the defeinition of those function to `with_match(op: Op, row: &RowRef<'_>, matched_row: &JoinRow<OwnedRow>)`

File: src/stream/src/executor/asof_join.rs [hunk 6]
```diff
@@ -0,0 +1,1379 @@
...
+use self::builder::JoinChunkBuilder;
+use super::barrier_align::*;
+use super::join::hash_join::*;
+use super::join::*;
+use super::watermark::*;
+use crate::executor::join::builder::JoinStreamChunkBuilder;
+use crate::executor::prelude::*;
+
+/// Evict the cache every n rows.
+const EVICT_EVERY_N_ROWS: u32 = 16;
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/asof_join.rs:36]: The should be different variables to control the different eviction frequencies. Consider they will only be used inside a file, using the same name is totally okay

File: proto/stream_plan.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuhao-su) [reply proto/stream_plan.proto:480]: > inequality key will be the first field in deduped_input_pk_indice

This is not necessary.

File: src/stream/src/executor/asof_join.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/asof_join.rs:938]: Will use `JoinType` instead of AsOfJoinType in proto in planner PR, so I can do the check then

**Review by st1page — APPROVED** [review]

**Pull request merged**